sui-execution/latest/sui-move-natives/src/transaction_context/mod.rs

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use better_any::{Tid, TidAble};
+use move_core_types::account_address::AccountAddress;
+use std::{cell::RefCell, rc::Rc};
+use sui_types::{
+    base_types::{ObjectID, SuiAddress, TxContext},
+    committee::EpochId,
+    digests::TransactionDigest,
+};
+
+// TransactionContext is a wrapper around TxContext that is exposed to NativeContextExtensions
+// in order to provide transaction context information to Move native functions.
+// Holds a Rc<RefCell<TxContext>> to allow for mutation of the TxContext.
+#[derive(Tid)]
+pub struct TransactionContext {
+    pub(crate) tx_context: Rc<RefCell<TxContext>>,
+    test_only: bool,
+}
+
+impl TransactionContext {
+    pub fn new(tx_context: Rc<RefCell<TxContext>>) -> Self {
+        Self {
+            tx_context,
+            test_only: false,
+        }
+    }
+
+    pub fn new_for_testing(tx_context: Rc<RefCell<TxContext>>) -> Self {
+        Self {
+            tx_context,
+            test_only: true,
+        }
+    }
+
+    pub fn sender(&self) -> SuiAddress {
+        self.tx_context.borrow().sender()
+    }
+
+    pub fn epoch(&self) -> EpochId {
+        self.tx_context.borrow().epoch()
+    }
+
+    pub fn epoch_timestamp_ms(&self) -> u64 {
+        self.tx_context.borrow().epoch_timestamp_ms()
+    }
+
+    pub fn digest(&self) -> TransactionDigest {
+        self.tx_context.borrow().digest()
+    }
+
+    pub fn sponsor(&self) -> Option<SuiAddress> {
+        self.tx_context.borrow().sponsor()
+    }
+
+    pub fn gas_price(&self) -> u64 {
+        self.tx_context.borrow().gas_price()
+    }
+
+    pub fn gas_budget(&self) -> u64 {
+        self.tx_context.borrow().gas_budget()
+    }
+
+    pub fn ids_created(&self) -> u64 {
+        self.tx_context.borrow().ids_created()
+    }
+
+    pub fn fresh_id(&self) -> ObjectID {
+        self.tx_context.borrow_mut().fresh_id()
+    }
+
+    //
+    // Test only function
+    //
+    pub fn replace(
+        &self,
+        sender: AccountAddress,
+        tx_hash: Vec<u8>,
+        epoch: u64,
+        epoch_timestamp_ms: u64,
+        ids_created: u64,
+    ) {
+        if !self.test_only {
```

Comment:
I'd _maybe_ consider an invariant violation/Result here instead. As the panic makes me a bit nervous

Reply:
I can change that. I did that just because it is so wrong to get through that path that I was not clear I cared enough

Reply:
changed, take a look and see if you like it more. @tzakian also wanted that I think.
I am not really sure what is the right thing and I keep going back and forth so take a look